File: backend-rust/src/format.rs
use serde_json::Value;

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum WireFormat {
    Json,
    FlatBuffers,
}

impl WireFormat {
    pub fn from_str(s: &str) -> Self {
        match s {
            "flatbuffers" | "fb" | "flatbuffer" => Self::FlatBuffers,
            _ => Self::Json,
        }
    }

    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Json => "json",
            Self::FlatBuffers => "flatbuffers",
        }
    }
}

pub fn encode_event(event: &str, data: &Value, _format: &WireFormat) -> Vec<u8> {
    let payload = serde_json::json!({"event": event, "data": data});
    serde_json::to_vec(&payload).unwrap_or_default()
}

pub fn decode_event(buf: &[u8]) -> (String, Value) {
    if let Ok(parsed) = serde_json::from_slice::<Value>(buf) {
        let event = parsed.get("event").and_then(|v| v.as_str()).unwrap_or("").to_string();
        let data = parsed.get("data").cloned().unwrap_or(Value::Null);
        return (event, data);
    }
    (String::new(), Value::Null)
}

pub fn encode_valkey(channel: &str, target: Option<&str>, event: &str, data: &Value, _format: &WireFormat) -> Vec<u8> {
    let payload = serde_json::json!({
        "channel": channel,
        "format": "json",
        "target": target,
        "event": event,
        "data": data,
    });
    serde_json::to_vec(&payload).unwrap_or_default()
}

pub fn decode_valkey(buf: &[u8]) -> (String, Option<String>, String, Value) {
    if let Ok(parsed) = serde_json::from_slice::<Value>(buf) {
        let channel = parsed.get("channel").and_then(|v| v.as_str()).unwrap_or("").to_string();
        let target = parsed.get("target").and_then(|v| v.as_str()).map(|s| s.to_string());
        let event = parsed.get("event").and_then(|v| v.as_str()).unwrap_or("").to_string();
        let data = parsed.get("data").cloned().unwrap_or(Value::Null);
        return (channel, target, event, data);
    }
    (String::new(), None, String::new(), Value::Null)
}

File: backend-node/format.js
const flatbuffers = require('flatbuffers');

const FORMAT_JSON = 'json';
const FORMAT_FLATBUFFERS = 'flatbuffers';

function detectFormat(auth = {}) {
  const f = (auth && auth.format) || 'json';
  return f === 'flatbuffers' || f === 'fb' ? FORMAT_FLATBUFFERS : FORMAT_JSON;
}

// --- SocketPayload manual FB builder/parser ---
const SP_SCHEMA = {
  fields: [
    { name: 'event', id: 0, vt: 4 },
    { name: 'data', id: 1, vt: 6 },
  ]
};

function buildSocketPayload(builder, event, dataBytes) {
  builder.startObject(2);
  const eventOff = builder.createString(event);
  builder.addFieldOffset(SP_SCHEMA.fields[0].vt, eventOff, 0);
  const dataOff = builder.createByteVector(dataBytes);
  builder.addFieldOffset(SP_SCHEMA.fields[1].vt, dataOff, 0);
  return builder.endObject();
}

function parseSocketPayload(buf) {
  const bb = new flatbuffers.ByteBuffer(new Uint8Array(buf));
  const root = bb.readInt32(bb.position()) + bb.position();
  const event = bb.__string(root + bb.__offset(root, SP_SCHEMA.fields[0].vt));
  const dataOff = bb.__offset(root, SP_SCHEMA.fields[1].vt);
  let data = null;
  if (dataOff) {
    const vecOff = root + dataOff;
    const len = bb.readInt32(vecOff);
    const raw = Buffer.from(bb.bytes().slice(vecOff + 4, vecOff + 4 + len));
    data = raw.toString('utf-8');
  }
  return { event, data };
}

// --- ValkeyEnvelope manual FB builder/parser ---
const VE_SCHEMA = {
  fields: [
    { name: 'channel', id: 0, vt: 4 },
    { name: 'format', id: 1, vt: 6 },
    { name: 'target', id: 2, vt: 8 },
    { name: 'event', id: 3, vt: 10 },
    { name: 'data', id: 4, vt: 12 },
  ]
};

function buildValkeyEnvelope(builder, channel, format, target, event, dataBytes) {
  builder.startObject(5);
  builder.addFieldOffset(VE_SCHEMA.fields[0].vt, builder.createString(channel), 0);
  builder.addFieldOffset(VE_SCHEMA.fields[1].vt, builder.createString(format), 0);
  if (target) builder.addFieldOffset(VE_SCHEMA.fields[2].vt, builder.createString(target), 0);
  builder.addFieldOffset(VE_SCHEMA.fields[3].vt, builder.createString(event), 0);
  if (dataBytes) builder.addFieldOffset(VE_SCHEMA.fields[4].vt, builder.createByteVector(dataBytes), 0);
  return builder.endObject();
}

function parseValkeyEnvelope(buf) {
  const bb = new flatbuffers.ByteBuffer(new Uint8Array(buf));
  const root = bb.readInt32(bb.position()) + bb.position();
  const channel = bb.__string(root + bb.__offset(root, VE_SCHEMA.fields[0].vt));
  const fmt = bb.__string(root + bb.__offset(root, VE_SCHEMA.fields[1].vt));
  const targetOff = bb.__offset(root, VE_SCHEMA.fields[2].vt);
  const target = targetOff ? bb.__string(root + targetOff) : null;
  const event = bb.__string(root + bb.__offset(root, VE_SCHEMA.fields[3].vt));
  const dataOff = bb.__offset(root, VE_SCHEMA.fields[4].vt);
  let data = null;
  if (dataOff) {
    const vecOff = root + dataOff;
    const len = bb.readInt32(vecOff);
    const raw = Buffer.from(bb.bytes().slice(vecOff + 4, vecOff + 4 + len));
    data = raw.toString('utf-8');
  }
  return { channel, format: fmt, target, event, data };
}

// --- Public API ---

function encodeEvent(event, data, format = FORMAT_JSON) {
  if (format === FORMAT_FLATBUFFERS) {
    const builder = new flatbuffers.Builder(1024);
    const dataStr = JSON.stringify(data);
    const payload = buildSocketPayload(builder, event, Buffer.from(dataStr, 'utf-8'));
    builder.finish(payload);
    return Buffer.from(builder.asUint8Array());
  }
  return Buffer.from(JSON.stringify({ event, data }));
}

function decodeEvent(buf) {
  try {
    const { event, data } = parseSocketPayload(buf);
    if (event !== null && data !== null) {
      return { event, data: JSON.parse(data) };
    }
  } catch (_) { /* fall through */ }

  try {
    const parsed = JSON.parse(buf.toString());
    return { event: parsed.event || '', data: parsed.data || null };
  } catch (_) {
    return { event: '', data: null };
  }
}

function encodeValkey(channel, target, event, data, format = FORMAT_JSON) {
  if (format === FORMAT_FLATBUFFERS) {
    const builder = new flatbuffers.Builder(1024);
    const dataStr = JSON.stringify(data);
    const dataBytes = dataStr ? Buffer.from(dataStr, 'utf-8') : null;
    const payload = buildValkeyEnvelope(builder, channel, FORMAT_FLATBUFFERS, target || null, event, dataBytes);
    builder.finish(payload);
    return Buffer.from(builder.asUint8Array());
  }
  return Buffer.from(JSON.stringify({ channel, format: FORMAT_JSON, target, event, data }));
}

function decodeValkey(buf) {
  try {
    const parsed = parseValkeyEnvelope(buf);
    if (parsed.channel !== null) {
      let data = null;
      if (parsed.data) {
        try { data = JSON.parse(parsed.data); } catch (_) { data = null; }
      }
      return { channel: parsed.channel, target: parsed.target, event: parsed.event, data };
    }
  } catch (_) { /* fall through */ }

  try {
    const parsed = JSON.parse(buf.toString());
    return {
      channel: parsed.channel || '',
      target: parsed.target || null,
      event: parsed.event || '',
      data: parsed.data || null,
    };
  } catch (_) {
    return { channel: '', target: null, event: '', data: null };
  }
}

module.exports = {
  FORMAT_JSON,
  FORMAT_FLATBUFFERS,
  detectFormat,
  encodeEvent,
  decodeEvent,
  encodeValkey,
  decodeValkey,
};
